peermanager/peerwantmanager.go

```diff
@@ -0,0 +1,226 @@
+package peermanager
+
+import (
+	"bytes"
+	"fmt"
+
+	lu "github.com/ipfs/go-bitswap/logutil"
+
+	cid "github.com/ipfs/go-cid"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+)
+
+type Gauge interface {
+	Inc()
+	Dec()
+}
+
+// peerWantManager
+type peerWantManager struct {
+	peerWants      map[peer.ID]*peerWant
+	wantBlockGauge Gauge
+}
+
+type peerWant struct {
+	wantBlocks *cid.Set
+	wantHaves  *cid.Set
+}
+
+// New creates a new peerWantManager.
+func newPeerWantManager(wantBlockGauge Gauge) *peerWantManager {
+	return &peerWantManager{
+		peerWants:      make(map[peer.ID]*peerWant),
+		wantBlockGauge: wantBlockGauge,
+	}
+}
+
+func (pwm *peerWantManager) AddPeer(p peer.ID) {
+	if _, ok := pwm.peerWants[p]; !ok {
+		pwm.peerWants[p] = &peerWant{
+			wantBlocks: cid.NewSet(),
+			wantHaves:  cid.NewSet(),
+		}
+	}
+}
+
+func (pwm *peerWantManager) RemovePeer(p peer.ID) {
+	delete(pwm.peerWants, p)
+}
+
+func (pwm *peerWantManager) BroadcastWantHaves(wantHaves []cid.Cid) map[peer.ID][]cid.Cid {
+	res := make(map[peer.ID][]cid.Cid)
+
+	// Iterate over all known peers
+	for p, pws := range pwm.peerWants {
+		// Iterate over all want-haves
+		for _, c := range wantHaves {
+			// If the CID has not been sent as a want-block or want-have
+			if !pws.wantBlocks.Has(c) && !pws.wantHaves.Has(c) {
+				// Record that the CID has been sent as a want-have
+				pws.wantHaves.Add(c)
+
+				// Add the CID to the results
+				if _, ok := res[p]; !ok {
+					res[p] = make([]cid.Cid, 0, 1)
+				}
+				res[p] = append(res[p], c)
+			}
+		}
+	}
+
+	return res
+}
+
+func (pwm *peerWantManager) SendWants(p peer.ID, wantBlocks []cid.Cid, wantHaves []cid.Cid) ([]cid.Cid, []cid.Cid) {
```

Comment:
The naming of these functions is a bit awkward. We aren't actually _sending_ anything. Maybe `PrepareWants`, `PrepareCancels`, or something like that?